step-plans/step-plans-simple-yaml-parser/src/main/java/step/plans/simple/schema/SimplePlanJsonSchemaGenerator.java

```diff
@@ -0,0 +1,452 @@
+	 *  },
+	 *  {
+	 *    "type": "object",
+	 *    "properties": {
+	 *      "selectionCriteria": {
+	 *        "$ref" : "#/$defs/DynamicKeywordInputsDef"
+	 *      }
+	 *    },
+	 *    "additionalProperties": false
+	 *  }
+	 *]
+	 */
+	private JsonObjectBuilder createCallKeywordFunctionNameDef() {
+		JsonObjectBuilder result = jsonProvider.createObjectBuilder();
+
+		JsonObjectBuilder keywordFunctionName = addRef(jsonProvider.createObjectBuilder(), SimpleDynamicValueJsonSchemaHelper.SMART_DYNAMIC_VALUE_STRING_DEF);
+
+		JsonObjectBuilder keywordSelectionCriteria = jsonProvider.createObjectBuilder();
+		keywordSelectionCriteria.add("type", "object");
+		keywordSelectionCriteria.add("properties", jsonProvider.createObjectBuilder()
+				.add("selectionCriteria", addRef(jsonProvider.createObjectBuilder(), SimpleDynamicValueJsonSchemaHelper.DYNAMIC_KEYWORD_INPUTS_DEF))
+		);
+		keywordSelectionCriteria.add("additionalProperties", false);
+
+		result.add("oneOf", jsonProvider.createArrayBuilder().add(keywordFunctionName).add(keywordSelectionCriteria));
+		return result;
+	}
+
+	private ArtefactDefinitions createArtefactImplDefs() throws JsonSchemaPreparationException {
+		ArtefactDefinitions artefactDefinitions = new ArtefactDefinitions();
+
+		// scan all @Artefact classes in classpath and automatically prepare definitions for them
+		Reflections r = new Reflections( new ConfigurationBuilder().forPackage(targetPackage));
+
+		// exclude artefacts from test packages
+		List<Class<?>> artefactClasses = r.getTypesAnnotatedWith(Artefact.class)
+				.stream()
+				.filter(c -> !c.getAnnotation(Artefact.class).test())
+				.collect(Collectors.toList());
+		log.info("The following {} artefact classes detected: {}", artefactClasses.size(), artefactClasses);
+
+		for (Class<?> artefactClass : artefactClasses) {
+			// use the name of artefact as definition name
+			String name = getArtefactName(artefactClass);
+			String defName = name + "Def";
+
+			// scan all fields in artefact class and put them to artefact definition
+			JsonObjectBuilder def = createArtefactImplDef(name, artefactClass);
+			artefactDefinitions.allArtefactDefs.put(defName, def);
+
+			// for root artefacts we only support the subset of all artefact definitions
+			for (RootArtefactType rootArtefactType : RootArtefactType.values()) {
+				if(rootArtefactType.createRootArtefact().getClass().equals(artefactClass)){
+					artefactDefinitions.rootArtefactDefs.add(defName);
+				}
+			}
+		}
+		return artefactDefinitions;
+	}
+
+	private static String getArtefactName(Class<?> artefactClass) {
+		Artefact ann = artefactClass.getAnnotation(Artefact.class);
+		return ann.name() == null || ann.name().isEmpty() ? artefactClass.getSimpleName() : ann.name();
+	}
+
+	private JsonObjectBuilder createArtefactImplDef(String name, Class<?> artefactClass) throws JsonSchemaPreparationException {
+		JsonObjectBuilder res = jsonProvider.createObjectBuilder();
+		res.add("type", "object");
```

Comment:
@iegorov777 Could we add a description field to the Artefact annotation and add it to the schema in order to have some inlined documentation of the Artefact?

Reply:
@jeromecomte Yes I think we could do so. 
Also it could be useful to add somehow the information about required fields for Artefact (now we generate the json schema without `required` limitations in artefacts). 
And also maybe add some annotations to define the yaml field name in case the field name in java differs from yaml format.